pkg/health/connection/connection.go

```diff
@@ -0,0 +1,87 @@
+/*
+Copyright (c) 2022 Nordix Foundation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package connection
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/nordix/meridio/pkg/health"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+)
+
+type gRPCConnectionStateMonitor struct {
+	healthService string
+	*grpc.ClientConn
+}
+
+// Monitor -
+// Monitors connection state changes and updates respective health service through context.
+// Currently only gRPC connection is supported, which relies on EXPERIMENTAL API.
+func Monitor(ctx context.Context, healthService string, cc interface{}) error {
+	switch cc := cc.(type) {
+	case *grpc.ClientConn:
+		m := gRPCConnectionStateMonitor{
+			healthService: healthService,
+			ClientConn:    cc,
+		}
+		go func() {
+			defer logrus.Debugf("Connection monitor exit (%s)", m.healthService)
+			for {
+				s := m.GetState()
+				if s == connectivity.Ready {
+					health.SetServingStatus(ctx, m.healthService, true)
+
+				} else {
+					health.SetServingStatus(ctx, m.healthService, false)
+				}
+				logrus.Tracef("Connection (%s) state %s", m.healthService, s)
+
+				// Note: gRPC will NOT establish undelying transport connection unless the user
+				// tries to send sg or it's initial connect (dial).
+				// Therefore trigger transport connect if state is Idle for 3 seconds, to avoid
+				// the health service from getting stuck in NOT_SERVING if the user "remains silent"
+				// for too long.
+				waitCtx := ctx
+				if s == connectivity.Idle {
+					// TODO: configurable timeout
+					cancelCtx, cancel := context.WithTimeout(ctx, 3*time.Second)
+					waitCtx = cancelCtx
+					defer cancel()
+				}
+				// Block until connection state changes.
+				if !m.WaitForStateChange(waitCtx, s) {
+					// context got timeout or canceled
+					select {
+					case <-ctx.Done():
+						// main context done
+						return
+					default:
+						// timeout; try re-connect
+						m.Connect()
```

Comment:
This is already done automatically by GRPC no? Also, not sure if it is the right place to have this, since this pkg only advertise the state of the connection and not manipulating the connection, what do you think?

Reply:
Unfortunately gRPC won't automatically "re-connect" for example after the server has disappeared temporarily.

This was dominant when I tested IPAM failover and noticed that readiness probe of proxies remained NOT_SERVING, until I forced some communication towards IPAM (e.g. by starting new targets) or simply restarted the proxy.

So even though I agree that this code should normally only monitor things, the whole feature wouldn't work properly with RPC protocols like IPAM which can be silent for long time and where there's no need for a stream to be opened all time (like watchers in case of NSP).

Reply:
But for instance if I delete the NSP pod, the load-balancer will be disconnected from the nsp service and will re-connect to it after the new pod is ready even without this `Connect()` call. Is it the GRPC state which has a bug or?

Reply:
In case of NSP it will reconnect, because the lb wants to maintain a watcher so basically a stream all the time, which involves sending things on the gRPC connection. And like I said, gRPC try to set up a transport connection if none available upon a "send request" from the user.
So with gRPC you will get a transport connection on your initial Dial and any subsequent sends if there's none for whatever reason.